fms/models/roberta.py

```diff
@@ -0,0 +1,331 @@
+
+        super(RoBERTa, self).__init__()
+        if config is not None:
+            self.config = config
+        else:
+            self.config = RoBERTaConfig()
+        self.config = self.config.updated(**kwargs)
+
+        self.base_model = RoBERTaHeadless(self.config)
+
+        self.class_head = ClassHead(
+            self.config.emb_dim,
+            str_to_activation(self.config.activation_fn),
+            self.config.norm_eps,
+        )
+
+        self.head = nn.Linear(self.config.emb_dim, self.config.src_vocab_size)
+
+        # this model ties weights, so we tie here
+        self.head.weight = self.base_model.embedding.weight
+
+    def forward(
+        self,
+        x: torch.LongTensor,
+        mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+    ):
+        # run through the encoder layers
+        x = self.base_model(x, mask=mask, position_ids=position_ids)
+
+        # run through the class head (using the first in each sequence in the batch as the cls_token)
+        x = self.class_head(x)
+
+        # project to vocab space
+        x = self.head(x)
+        return x
+
+    @classmethod
+    def from_config(cls, config: RoBERTaConfig) -> "RoBERTa":
+        return cls(config)
+
+    def get_config(self) -> RoBERTaConfig:
+        return self.config
+
+    def reset_params(self):
+        # Modules are self-initializing, we're just going to down-scale the final prediction head to be
+        # mixed-fan (inputs and gradients scale to the same inverse factors) if it isn't tied
+        self.head.weight.data.normal_(
+            0,
+            1 / math.sqrt(math.sqrt(self.config.emb_dim * self.config.src_vocab_size)),
+        )
+
+
+# a micro llama model to use with a char-level tokenizer
+_micro_char_config = RoBERTaConfig(
+    emb_dim=192, nheads=4, nlayers=5, max_pos=1024, src_vocab_size=256
+)
+
+_base_config = RoBERTaConfig()
+
+_architecture_name = "roberta"
+
+
+def _roberta_factory_factory(config):
+    def factory(**kwargs):
+        return RoBERTa(config, **kwargs)
+
+    return factory
+
+
+models.register_model(
+    _architecture_name, "micro", _roberta_factory_factory(_micro_char_config)
+)
+models.register_model(
+    _architecture_name, "base", _roberta_factory_factory(_base_config)
+)
+
+
+def _hf_sd_to_fms_sd(hf_sd):
+    result = {}
+
+    # process embeddings
+    result["base_model.embedding.weight"] = hf_sd[
+        "roberta.embeddings.word_embeddings.weight"
+    ]
+    result["base_model.position_embedding.weight"] = hf_sd[
+        "roberta.embeddings.position_embeddings.weight"
+    ][2:]
+
+    def _apply_weight_bias(hf_value, fms_value):
+        result[f"{fms_value}.weight"] = hf_sd[f"{hf_value}.weight"]
+        result[f"{fms_value}.bias"] = hf_sd[f"{hf_value}.bias"]
+
+    # process layers
+    layer_pattern = re.compile("roberta.encoder.layer.[0-9]+")
+    processed_layers = set()
+    for hf_k, hf_v in hf_sd.items():
```

Comment:
it looks like this is doing redundant copies?

i.e. iterates through everything in the sd that matches `layer` ... and for each row it looks up everything in the sd for that layer... so if it finds, say, layer.1.key and layer.1.query ... it's going to copy key + query when it finds layer.1.key and thne copy key and query again when it finds layer.1.query?

or am I reading this wrong?

Also, I think this might not work properly if the sd gets arbitrarily split into layer shards (which may not always shard on grouped "layer" boundaries)

Reply:
No, at most this will just see a layer and skip it. A few lines down, I am doing:

```python
# only process the layer if we have not seen it yet
            if layer not in processed_layers:
```

Basically this is just matching on a layer with `layer_pattern`. Then doing everything for the layer (but skipping layers it has already seen)

Reply:
hmm... I think we could write this in a simpler way, and also make it safer for layer-sharded loading?
Ideally we should be able to adapt any individual key+tensor pair.